.github/workflows/build.yml

```diff
@@ -3,33 +3,90 @@ name: Build
 on: push
 
 jobs:
-  pytest:
+  tests:
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version:
+        - "3.6"
+        - "3.7"
+        - "3.8"
+        - "3.9"
+        - "3.10"
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v2
-      - name: Set up Python 3.8.6
+
+      - name: Set up Python ${{ matrix.python-version }}
         uses: actions/setup-python@v2
         with:
-          python-version: 3.8.6
-      - name: Install poetry
-        run: python -m pip install --upgrade pip poetry
+          python-version: ${{ matrix.python-version }}
+
+      - name: Get pip cache dir
+        id: pip-cache
+        run: |
+          echo "::set-output name=dir::$(pip cache dir)"
+
+      - name: Cache pip
+        uses: actions/cache@v2
+        with:
+          path: ${{ steps.pip-cache.outputs.dir }}
+          key:
+            ${{ matrix.python-version }}-v1-${{ hashFiles('**/pyproject.toml') }}-${{ hashFiles('**/tox.ini') }}
+          restore-keys: |
+            ${{ matrix.python-version }}-v1-
+
+      - name: Install Tox
+        run: python -m pip install --upgrade pip tox tox-gh-actions
+
+      - name: Run tox
+        run: tox
+
+      - name: Upload coverage data
+        uses: actions/upload-artifact@v2
+        with:
+          name: coverage-data
+          path: '.coverage.*'
+
+  coverage:
+    runs-on: ubuntu-latest
+    needs: tests
+    steps:
+      - uses: actions/checkout@v2
+
+      - uses: actions/setup-python@v2
+        with:
+          python-version: '3.10'
+
       - name: Install dependencies
-        run: poetry install
-      - name: Run pytest
-        run: poetry run pytest --cov --cov-report xml
-      - name: Upload coverage to Codecov
-        uses: codecov/codecov-action@v1
+        run: python -m pip install --upgrade coverage[toml]
+
+      - name: Download data
+        uses: actions/download-artifact@v2
         with:
-          token: ${{ secrets.CODECOV_TOKEN }}
-          fail_ci_if_error: true
+          name: coverage-data
+
+      - name: Combine coverage and fail if it's <100%
+        run: |
+          python -m coverage combine
+          python -m coverage html --skip-covered --skip-empty
+          python -m coverage report --fail-under=100
+
+      - name: Upload HTML report
+        if: ${{ failure() }}
+        uses: actions/upload-artifact@v2
+        with:
+          name: html-report
+          path: htmlcov
+
   black:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v2
-      - name: Set up Python 3.8.6
+      - name: Set up Python 3.9.9
         uses: actions/setup-python@v2
         with:
-          python-version: 3.8.6
+          python-version: 3.9.9
```

Comment:
Why 3.9? It works, but if 3.10 works, that seems preferable.

Reply:
Good point.

Reply:
Good point. - take two